src/qibolab/instruments/qm.py

```diff
@@ -963,17 +1025,44 @@ def sweep_start(self, sweepers, qubits, qmsequence, relaxation_time):
                 to_process = {qmpulse}
                 while to_process:
                     next_qmpulse = to_process.pop()
-                    to_process |= next_qmpulse.next
+                    to_process |= next_qmpulse.next_pulses
                     next_qmpulse.wait_time_variable = start
 
             self.sweep_recursion(sweepers[1:], qubits, qmsequence, relaxation_time)
 
+    def sweep_duration(self, sweepers, qubits, qmsequence, relaxation_time):
+        sweeper = sweepers[0]
+        for pulse in sweeper.pulses:
+            qmpulse = qmsequence.pulse_to_qmpulse[pulse.serial]
+            if isinstance(qmpulse, BakedPulse):
+                values = np.array(sweeper.values).astype(int)
+                qmpulse.bake(self.config, values)
+            else:
+                values = np.array(sweeper.values).astype(int) // 4
+
+        dur = declare(int)
+
+        with for_(*from_array(dur, values)):
```

Comment:
Indeed, that was both right and wrong because `values` stay the same during the loop (the `Sweeper` has a list of pulses but the `values` are the same array for all pulses). I update it so that the `values` calculation is outside the loop which is conceptually better. Thanks for catching this!